docs/isl-2-0/spec.md

```diff
@@ -0,0 +1,673 @@
+For numeric values, a value is valid if and only if it can be losslessly converted from the Ion representation to the specified float representation, and back to the Ion representation.
+
+{% comment %}{% include example.md title="`ieee754_float`" code_file="examples/placeholder.isl" %}{% endcomment %}
+
+## not
+
+{% include grammar-element.md productions="not" %}
+
+Value must not be valid for the type.
+
+{% comment %}{% include example.md title="`not`" code_file="examples/placeholder.isl" %}{% endcomment %}
+
+## one_of
+
+{% include grammar-element.md productions="one_of" %}
+
+Value must be valid for exactly one of the types. The list of types must not be empty.
+
+{% include note.html type="important" content="
+Specifying two overlapping types may cause unexpected results.
+For example, if you specify `one_of: [$null_or::int, $null_or::float]`, the value `null` will be invalid because it matches two type references. 
+" %}
+
+{% comment %}{% include example.md title="`one_of`" code_file="examples/placeholder.isl" %}{% endcomment %}
+
+## ordered_elements
+
+{% include grammar-element.md productions="ordered_elements" %}
+
+Defines constraints over a list of values in a heterogeneous list, S-expression, or document.
+Each value in a list, S-expression, or document is expected to be valid against the type in the corresponding position of the specified types list.
+Each type is implicitly defined with `occurs: 1`—behavior which may be overridden.
+(See [§Variably-Occurring Type References](#variably-occurring-type-references).)
+
+When specified, this constraint fully defines the content of a list, S-expression, or document—open content is not allowed.
+
+{% comment %}{% include example.md title="`ordered_elements`" code_file="examples/placeholder.isl" %}{% endcomment %}
+
+## precision
+
+{% include grammar-element.md productions="precision" %}
+
+Applicable for `decimal`.
+An exact or minimum/maximum indicating the number of digits in the unscaled value of a decimal.
+The minimum precision must be greater than or equal to `1`.
+
+{% comment %}{% include example.md title="`precision`" code_file="examples/placeholder.isl" %}{% endcomment %}
+
+## regex
+
+{% include grammar-element.md productions="regex" %}
+
+Tests whether a `symbol` or `string` has any matches for the given regular expression.
+The regular expression is a string that conforms to a *RegularExpressionBody* defined by [ECMA 262 Regular Expressions](https://www.ecma-international.org/ecma-262/5.1/#sec-7.8.5).
+Regular expressions shall be limited to the following features:
+
+|                     |                                      |
+|--------------------:|--------------------------------------|
+|                     | Unicode codepoints match themselves  |
+|                 `.` | any codepoint                        |
+|             `[abc]` | codepoint class                      |
+|             `[a-z]` | range codepoint class                |
```

Comment:
I think basic regex supports a combined class like `[a-nz]` (the range `a-n` union `z`). Do we support that?

Reply:
Yes. We also support `[a-f\d]`. I'll update the table to reflect these.